### PR TITLE
[ticket/14557] Simplify updating overloaded events for extensions

### PR DESCRIPTION
**Pull request by senky**
Checklist:

- [ ] Correct branch: master for new features; 3.2.x, 3.1.x for fixes
- [ ] Tests pass
- [ ] Code follows coding guidelines: [master / 3.2.x](https://area51.phpbb.com/docs/master/coding-guidelines.html), [3.1.x](https://area51.phpbb.com/docs/31x/coding-guidelines.html)
- [ ] Commit follows commit message [format](https://wiki.phpbb.com/Git#Commit_Messages)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-14557

PHPBB3-14557

### CODE REVIEW

**Comment by bantu** [comment]
I am not sure this is such a good idea. It seems to be inherently limited to one-dimensional arrays. Is there a solution that is not limited with regards to that?

**Comment by senky** [comment]
You can always use `offsetSet()` to replace whole array. It's just helper. To make life easier at least a little bit...

**Comment by Nicofuma** [comment]
I'm not against a such DX improvement, but it needs another name and a docblock imho (a name stating what it does).

**Comment by senky** [comment]
`offsetKeyUpdate`?

**Comment by 3D-I** [comment]
offsetKeyUpdate looks fine to me, but that's me.
btw only one function in this file got to have the docblock,

**Comment by Nicofuma** [comment]
yes, the others are either explicit (set_data) either magic methods from `\ArrayAccess` but they could have a docblock too. About the name I'd prefer a really explicit "phpBB" name, not a name looking like it's part of \ArrayAccess something like update_subarray or something like that. At the same time you will need to add some error control: `$this->data[$offset]` may not exist or not be an array

**Comment by senky** [comment]
"phpBB" name and docblock added. I would rather not control any errors. If extension author tries to update subarray that is not array, it is always his bad, since event input is well documented and it's content is guaranteed.

**Comment by VSEphpbb** [comment]
If you're gonna call it subarray then all occurrences of offset should be replaced by subarray (the var name and in the docblocks)

**Comment by VSEphpbb** [comment]
Error control would be easy, something like:

```php
if (!is_array($this->data[$subarray])
{
    throw new \phpbb\exception\runtime_exception('INVALID_SUBARRAY', $subarray);
}
```
@Nicofuma ?

Even though it may ext author's bad, it helps if they can find out where they made a bad.

**Comment by bantu** [comment]
Why not https://github.com/igorw/get-in/blob/master/src/get_in.php#L28 ?

File: phpBB/phpbb/event/data.php [hunk 1]
```diff
@@ -63,4 +63,17 @@ public function offsetUnset($offset)
...
 		unset($this->data[$offset]);
 	}
+
+	/**
+	 * Returns data with updated key in specified offset.
+	 *
+	 * @param	string	$offset	Data array offset
+	 * @param	string	$key	Offset key
+	 * @param	mixed	$value	Value to update
+	 * @return	mixed	Value of $value
```
Comment (bantu) [inline phpBB/phpbb/event/data.php]: What is the purpose of this? Why is there a return value?

Comment (VSEphpbb) [inline phpBB/phpbb/event/data.php]: Read the ticket ;)

Comment (bantu) [inline phpBB/phpbb/event/data.php]: Ticket does not explain. Reason should be document in code anyway.

**Comment by senky** [comment]
@bantu do you mean to rewrite to the same principle as `update_in` in igorw/get-in? `callable` instead of direct value?

File: phpBB/phpbb/event/data.php (hunk 1 above)
Comment (senky) [inline phpBB/phpbb/event/data.php]: We either return nothing or `$value`. There is nothing more to that, I just followed what other methods in the class do.

**Comment by VSEphpbb** [comment]
This should be in the event/data class, as that is already available. Shouldn't come from some external class or library that needs to be added.

File: phpBB/phpbb/event/data.php (hunk 1 above)
Comment (VSEphpbb) [inline phpBB/phpbb/event/data.php]: It does not need to be returned @senky

**Comment by bantu** [comment]
@VSEphpbb I am pretty sure adding get-in as a dependency is an option. I however mostly posted the link to show that it is not too hard to come up with an implementation that is not limited to one-dimensional arrays (as mentioned in my first post to this PR).

**Comment by senky** [comment]
@bantu while adding get-in as a dependency is an option, adding new dependency should probably be discussed and approved. This ticket, however, just tries to make life a lot easier for simple tasks like replacing template variable. Maybe we could merge this with shallow subarray support and if need emerges in the future, we can refactor for multidimensional arrays.

**Comment by VSEphpbb** [comment]
Adding get-in as a dependency is not an option @senky As @bantu said he posted it as a coding example

**Comment by marc1706** [comment]
Bump @senky

**Comment by senky** [comment]
@marc1706 I am not sure what's missing here from the discussion. Subarray support? Using get-in?

**Comment by marc1706** [comment]
Was just wondering whether you wanted to change anything else since it seemed like the last comments kind of left things off.

**Pull request merged**